README.md
# trace.moe-worker

Workers for [trace.moe](https://github.com/soruly/trace.moe)

- upload script
- hasher
- loader


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [unreleased] - 2021-02-30


watch.js
import "dotenv/config.js";
import fetch from "node-fetch";
import fs from "fs-extra";
import path from "path";
import chokidar from "chokidar";

const { TRACE_API_URL, TRACE_API_SECRET, TRACE_MEDIA_URL, TRACE_WATCH_PATH } = process.env;

console.log(`Watching ${TRACE_WATCH_PATH} for new files`);
chokidar
  .watch(TRACE_WATCH_PATH, {
    persistent: true,
    ignoreInitial: false,
    usePolling: false,
    awaitWriteFinish: {
      stabilityThreshold: 2000,
      pollInterval: 100,
    },
    atomic: true, // or a custom 'atomicity delay', in milliseconds (default 100)
  })
  .on("add", async (filePath) => {
    console.log(`[chokidar] add ${filePath}`);
    if (!fs.existsSync(filePath)) {
      console.log(`Gone ${filePath}`);
      return;
    }
    if (filePath.replace(TRACE_WATCH_PATH, "").split("/").length < 2) return;

    const anilistID = filePath.replace(TRACE_WATCH_PATH, "").split("/")[0];
    const fileName = filePath.replace(TRACE_WATCH_PATH, "").split("/").pop();

    if (![".mp4"].includes(path.extname(fileName).toLowerCase())) {
      console.log(`Delete ${filePath}`);
      fs.removeSync(filePath);
      return;
    }

    console.log(`Uploading ${anilistID}/${fileName}`);
    const res = await fetch(
      `${TRACE_MEDIA_URL}/file/${anilistID}/${encodeURIComponent(fileName)}`,
      {
        method: "PUT",
        body: fs.createReadStream(filePath),
        headers: { "x-trace-secret": TRACE_API_SECRET },
      }
    );
    if (res.status === 201 || res.status === 204) {
      console.log(`Uploaded ${anilistID}/${fileName}`);
      const response = await fetch(
        `${TRACE_API_URL}/uploaded/${anilistID}/${encodeURIComponent(fileName)}`,
        {
          headers: { "x-trace-secret": TRACE_API_SECRET },
        }
      );
      if (response.status !== 204) {
        console.log(`Error: API update failed. HTTP ${response.status}`);
      } else {
        fs.removeSync(filePath);
        console.log("Completed");
      }
    } else {
      console.log(`Error: upload failed. HTTP ${res.status}`);
    }
  })
  .on("unlink", (filePath) => {
    console.log(`[chokidar] unlink ${filePath}`);
    if (fs.readdirSync(path.dirname(filePath)).length === 0) {
      console.log(`Removing ${path.dirname(filePath)}`);
      fs.removeSync(path.dirname(filePath));
    }
  })
  .on("unlinkDir", (dirPath) => {
    console.log(`[chokidar] unlinkDir ${dirPath}`);
    if (
      fs.readdirSync(path.dirname(dirPath)).length === 0 &&
      `${path.dirname(dirPath)}/` !== TRACE_WATCH_PATH
    ) {
      console.log(`Removing ${path.dirname(dirPath)}`);
      fs.removeSync(path.dirname(dirPath));
    }
  });


loader.js
import "dotenv/config.js";
import WebSocket from "ws";
import xmldoc from "xmldoc";
import lzma from "lzma-native";
import fetch from "node-fetch";

const { TRACE_API_URL, TRACE_API_SECRET } = process.env;

let ws;
const openHandle = () => {
  console.log("connected");
};

const messageHandle = async (message) => {
  const { file, core } = JSON.parse(message);

  console.log(`Downloading ${file}`);
  const [anilistID, fileName] = file.split("/");
  const res = await fetch(
    `${TRACE_API_URL}/hash/${anilistID}/${encodeURIComponent(fileName)}.xml.xz`,
    {
      headers: { "x-trace-secret": TRACE_API_SECRET },
    }
  );
  if (res.status >= 400) {
    console.log(`Error: Fail to download "${await res.text()}"`);
    ws.send(message);
    return;
  }

  console.log("Unzipping hash");
  const data = await lzma.decompress(await res.buffer());

  console.log("Parsing xml");
  const hashList = new xmldoc.XmlDocument(data).children
    .filter((child) => child.name === "doc")
    .map((doc) => {
      const fields = doc.children.filter((child) => child.name === "field");
      return {
        time: parseFloat(fields.filter((field) => field.attr.name === "id")[0].val),
        cl_hi: fields.filter((field) => field.attr.name === "cl_hi")[0].val,
        cl_ha: fields.filter((field) => field.attr.name === "cl_ha")[0].val,
      };
    })
    .sort((a, b) => a.time - b.time);

  const dedupedHashList = [];
  hashList.forEach((currentFrame) => {
    if (
      !dedupedHashList
        .slice(-24) // get last 24 frames
        .filter((frame) => currentFrame.time - frame.time < 2) // select only frames within 2 sec
        .some((frame) => frame.cl_hi === currentFrame.cl_hi) // check for exact match frames
    ) {
      dedupedHashList.push(currentFrame);
    }
  });

  const xml = [
    "<add>",
    dedupedHashList
      .map((doc) =>
        [
          "<doc>",
          '<field name="id">',
          `<![CDATA[${file}/${doc.time.toFixed(2)}]]>`,
          "</field>",
          '<field name="cl_hi">',
          doc.cl_hi,
          "</field>",
          '<field name="cl_ha">',
          doc.cl_ha,
          "</field>",
          "</doc>",
        ].join("")
      )
      .join("\n"),
    "</add>",
  ].join("\n");

  console.log(`Uploading xml to ${core}`);
  await fetch(`${core}/update?wt=json&commit=true`, {
    method: "POST",
    headers: { "Content-Type": "text/xml" },
    body: xml,
  });

  await fetch(`${TRACE_API_URL}/loaded/${anilistID}/${encodeURIComponent(fileName)}`, {
    headers: { "x-trace-secret": TRACE_API_SECRET },
  });
  ws.send(message);
  console.log(`Loaded ${file}`);
};

const closeHandle = async () => {
  console.log("connecting");
  ws = new WebSocket(`${TRACE_API_URL}/ws`, {
    headers: { "x-trace-secret": TRACE_API_SECRET, "x-trace-worker-type": "load" },
  });
  ws.on("open", openHandle);
  ws.on("message", messageHandle);
  ws.on("close", async () => {
    await new Promise((resolve) => setTimeout(resolve, 5000));
    closeHandle();
  });
};

closeHandle();


upload.js
import "dotenv/config.js";
import fetch from "node-fetch";
import fs from "fs-extra";

const { TRACE_API_URL, TRACE_API_SECRET, TRACE_MEDIA_URL } = process.env;

const fileToUpload = "/mnt/data/anime/1/[KANI][Charlotte][NCED][MP4][1920x1080].mp4";
const anilistID = "1";
const filename = "[KANI][Charlotte][NCED][MP4][1920x1080].mp4";

console.log(`Uploading ${fileToUpload}`);
const res = await fetch(`${TRACE_MEDIA_URL}/${anilistID}/${encodeURIComponent(filename)}`, {
  method: "PUT",
  body: fs.createReadStream(fileToUpload),
  headers: { "x-trace-secret": TRACE_API_SECRET },
});
if (res.status === 201 || res.status === 204) {
  console.log(`Uploaded ${fileToUpload}`);
  const response = await fetch(
    `${TRACE_API_URL}/uploaded/${anilistID}/${encodeURIComponent(filename)}`,
    {
      headers: { "x-trace-secret": TRACE_API_SECRET },
    }
  );
  if (response.status !== 204) {
    console.log(`Error: API update failed. HTTP ${response.status}`);
  } else {
    console.log("Completed");
  }
} else {
  console.log(`Error: upload failed. HTTP ${res.status}`);
}


hasher.js
import "dotenv/config.js";
import WebSocket from "ws";
import path from "path";
import os from "os";
import fs from "fs-extra";
import child_process from "child_process";
import lzma from "lzma-native";
import fetch from "node-fetch";

const { TRACE_API_URL, TRACE_API_SECRET, TRACE_MEDIA_URL } = process.env;

let ws;
const openHandle = () => {
  console.log("connected");
};

const messageHandle = async (message) => {
  const { file, algo } = JSON.parse(message);
  console.log(`Hashing ${file}`);

  const tempPath = path.join(os.tmpdir(), `sola-${process.pid}`);
  console.log(`Creating temp directory ${tempPath}`);
  fs.ensureDirSync(tempPath);
  fs.emptyDirSync(tempPath);

  const imageDescriptor = [
    "cl",
    "eh",
    "jc",
    "oh",
    "ph",
    "ac",
    "ad",
    "ce",
    "fc",
    "fo",
    "jh",
    "sc",
  ].includes(algo)
    ? algo
    : null;

  if (!imageDescriptor) {
    console.log(`Error: Unsupported image descriptor "${algo}"`);
    ws.send(message);
    return;
  }

  console.log(`Downloading ${file}`);
  const [anilistID, fileName] = file.split("/");
  const mp4FilePath = path.join(tempPath, "video.mp4");
  const video = await fetch(
    `${TRACE_MEDIA_URL}/file/${anilistID}/${encodeURIComponent(fileName)}`,
    {
      headers: { "x-trace-secret": TRACE_API_SECRET },
    }
  );
  if (video.status >= 400) {
    console.log(`Error: Fail to download video "${await video.text()}"`);
    ws.send(message);
    return;
  }
  await new Promise((resolve, reject) => {
    const fileStream = fs.createWriteStream(mp4FilePath);
    video.body.pipe(fileStream);
    video.body.on("error", (err) => {
      reject(err);
    });
    fileStream.on("finish", function () {
      resolve();
    });
  });
  if (!fs.existsSync(mp4FilePath)) {
    console.log("Error: Fail to download video");
    return;
  }

  console.log("Extracting thumbnails");
  const { stderr: ffmpegLog } = child_process.spawnSync(
    "ffmpeg",
    [
      "-i",
      mp4FilePath,
      "-q:v",
      2,
      "-an",
      "-vf",
      "fps=12,scale=-2:180,showinfo",
      `${tempPath}/%08d.jpg`,
    ],
    { encoding: "utf-8", maxBuffer: 1024 * 1024 * 100 }
  );
  fs.unlinkSync(mp4FilePath);
  const myRe = /pts_time:\s*((\d|\.)+?)\s*pos/g;
  let temp = [];
  const timeCodeList = [];
  while ((temp = myRe.exec(ffmpegLog)) !== null) {
    timeCodeList.push(parseFloat(temp[1]).toFixed(4));
  }
  console.log(`Extracted ${timeCodeList.length} timecode`);

  const thumbnailList = fs.readdirSync(tempPath);
  console.log(`Extracted ${thumbnailList.length} thumbnails`);

  console.log("Preparing frame files for analysis");
  const thumbnailListPath = path.join(tempPath, "frames.txt");
  fs.writeFileSync(
    thumbnailListPath,
    thumbnailList.map((each) => path.join(tempPath, each)).join("\n")
  );

  console.log("Analyzing frames");
  const lireSolrXMLPath = path.join(tempPath, "output.xml");
  const { stdout, stderr } = child_process.spawnSync(
    "java",
    [
      "-cp",
      "jar/*",
      "net.semanticmetadata.lire.solr.indexing.ParallelSolrIndexer",
      "-i",
      thumbnailListPath,
      "-o",
      lireSolrXMLPath,
      "-f", // force to overwrite output file
      // "-a", // use both BitSampling and MetricSpaces
      // "-l", // disable bitSampling and use MetricSpaces instead
      "-w", // monitoring interval, default 1000ms
      500,
      "-n", // number of threads
      16,
      "-y", // defines which feature classes are to be extracted, comma seperated
      imageDescriptor, // cl,eh,jc,oh,ph,ac,ad,ce,fc,fo,jh,sc
    ],
    { encoding: "utf-8", maxBuffer: 1024 * 1024 * 100 }
  );
  console.log(stdout);
  console.log(stderr);

  console.log("Post-Processing XML");
  // replace frame numbers with timecode
  // and sort by timecode in ascending order
  const parsedXML = [
    "<add>",
    fs
      .readFileSync(lireSolrXMLPath, "utf-8")
      .split("\n")
      .map((line) => line.trim())
      .filter((line) => line.indexOf("<doc>") === 0)
      .map((line) =>
        line
          .replace(/<field name="title">(.*?)<\/field>/g, "")
          .replace(
            /<field name="id">.*\/(.*?\.jpg)<\/field>/g,
            (match, p1) => `<field name="id">${timeCodeList[thumbnailList.indexOf(p1)]}</field>`
          )
      )
      .sort(
        (a, b) =>
          parseFloat(a.match(/<field name="id">(.*?)<\/field>/)[1]) -
          parseFloat(b.match(/<field name="id">(.*?)<\/field>/)[1])
      )
      .join("\n"),
    "</add>",
  ].join("\n");
  // fs.writeFileSync("debug.xml", parsedXML);
  console.log("Removing temp files");
  fs.removeSync(tempPath);

  console.log("Compressing XML");
  const compressedXML = await lzma.compress(parsedXML, { preset: 6 });

  console.log(`Uploading ${file}`);
  await fetch(`${TRACE_API_URL}/hash/${anilistID}/${encodeURIComponent(fileName)}`, {
    method: "PUT",
    body: compressedXML,
    headers: { "x-trace-secret": TRACE_API_SECRET },
  });
  ws.send(message);
  console.log(`Uploaded ${file}`);
};

const closeHandle = async () => {
  console.log("connecting");
  ws = new WebSocket(`${TRACE_API_URL}/ws`, {
    headers: { "x-trace-secret": TRACE_API_SECRET, "x-trace-worker-type": "hash" },
  });
  ws.on("open", openHandle);
  ws.on("message", messageHandle);
  ws.on("close", async () => {
    await new Promise((resolve) => setTimeout(resolve, 5000));
    closeHandle();
  });
};

closeHandle();
